### PR TITLE
feat: Corellium communication layer

### LINKED ISSUE #1799: Corellium communication layer
## Goal

* Design simple abstraction over the Corellium communication layer.
    * Authentication.
    * Running test on multiple devices.
* Implement proof of concept test for running multiple tests on multiple Corellium android devices.
* Print logs directly from running devices.

## Non-goal

* Sharding
* Advanced device management
    * Creating new.
    * Invoking if not running.
* Advanced log parsing

### PR DESCRIPTION
**Pull request by jan-gogo**
Fixes #1799 

## API

Declaration of interfaces and structures with java doc:
https://github.com/Flank/flank/tree/1799_Corellium_communication_layer/corellium/api/src/main/kotlin/flank/corellium/api

## Client

architecture diagram

.

## Checklist

- [x] API structures and interfaces
- [x] Refactor client from OOP to FP
- [x] Update sandbox for client v2
- [x] Implement adapters
- [x] Documented
- [x] Example main function

### CODE REVIEW

**Comment by bootstraponline** [comment]
the architecture diagram looks excellent!

File: corellium/client/src/main/kotlin/flank/corellium/client/agent/Connect.kt [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                    handleIncomingFrames()
+                    waitForReady()
+                }
+        } catch (ex: Exception) {
+            ex.printStackTrace()
+            delay(20_000)
+        }
+    } while (connection == null)
+
+    connection
```
Comment (piotradamczyk5) [inline corellium/client/src/main/kotlin/flank/corellium/client/agent/Connect.kt:50]: could you separate getting client and connection to separate function?

File: corellium/adapter/src/main/kotlin/flank/corellium/adapter/InstallAndroidApps.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import flank.corellium.api.AndroidApps
+import flank.corellium.client.agent.uploadFile
+import flank.corellium.client.console.sendCommand
+import flank.corellium.client.core.connectAgent
+import flank.corellium.client.core.connectConsole
+import flank.corellium.client.core.getAllProjects
+import flank.corellium.client.core.getProjectInstancesList
+import java.io.File
+
+private const val pathToUpload = "/sdcard"
```
Comment (Sloox) [inline corellium/adapter/src/main/kotlin/flank/corellium/adapter/InstallAndroidApps.kt:12]: Constants -> Capitalization?

File: corellium/adapter/src/test/kotlin/flank/corellium/adapter/AdaptersTest.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import flank.corellium.client.core.getAllProjects
+import flank.corellium.client.core.getProjectInstancesList
+import flank.corellium.corelliumApi
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.runBlocking
+
+private const val PROJECT_NAME = "Default Project"
+
+private val credentials = Authorization.Credentials(
+    host = TODO(),
```
Comment (Sloox) [inline corellium/adapter/src/test/kotlin/flank/corellium/adapter/AdaptersTest.kt:15]: TODOS in a PR? 
Can we link them to tickets numbers if its not meant to be here to keep clarity.

File: corellium/api/src/main/kotlin/flank/corellium/api/AndroidApps.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package flank.corellium.api
+
+/**
+ * The [AndroidApps] is representing a bunch of apk files related to the testing instance.
```
Comment (Sloox) [inline corellium/api/src/main/kotlin/flank/corellium/api/AndroidApps.kt:4]:
```suggestion
 * [AndroidApps] represents a bunch of apk files related to the testing instance.
```

File: corellium/client/src/main/kotlin/flank/corellium/client/agent/Connect.kt [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                .let(::Agent)
+                .apply {
+                    handleIncomingFrames()
+                    waitForReady()
+                }
+        } catch (ex: Exception) {
+            ex.printStackTrace()
+            delay(20_000)
+        }
+    } while (connection == null)
```
Comment (Sloox) [inline corellium/client/src/main/kotlin/flank/corellium/client/agent/Connect.kt:48]: possible infinite loop?

**Review by Sloox — CHANGES_REQUESTED** [review]
Minor changes.

File: corellium/adapter/src/test/kotlin/flank/corellium/adapter/AdaptersTest.kt (hunk 3 above)
Comment (jan-gogo) [reply corellium/adapter/src/test/kotlin/flank/corellium/adapter/AdaptersTest.kt:15]: It's ok at this stage of that work in progress MVP.

File: corellium/client/src/main/kotlin/flank/corellium/client/agent/Connect.kt (hunk 5 above)
Comment (jan-gogo) [reply corellium/client/src/main/kotlin/flank/corellium/client/agent/Connect.kt:50]: fixed by adding require at the beginning of `do`

**Pull request closed without merging**